### PR TITLE
SpectrumDetectorMapping constructor

### LINKED ISSUE #23607: SpectrumDetectorMapping constructor
One way of constructing a spectrum-detector-map is to pass a workspace pointer.
This constructor is only used in one place: `CopyDetectorMapping`.

It may be useful to use a `MatrixWorkspace_const_sptr` for the construction (getting the map from a workspace).

This change may be little and not risky.

Most of the maps are constructed differently and used in `DataHandling`.

### PR DESCRIPTION
**Pull request by vreimund**
**Description of work.**

This is a proposition of changing a constructor in `SpectrumDetectorMapping`.

Takes a MatrixWorkspace_const_sptr instead of const MatrixWorkspace
- Only usage is in `CopyDetecorMapping` algorithm

Fixes #23607.

*This does not require release notes*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/inc/MantidAPI/SpectrumDetectorMapping.h [hunk 1]
```diff
@@ -53,7 +47,7 @@ class MANTID_API_DLL SpectrumDetectorMapping {
   using sdmap = std::unordered_map<specnum_t, std::set<detid_t>>;
 
 public:
-  explicit SpectrumDetectorMapping(const MatrixWorkspace *const workspace,
+  explicit SpectrumDetectorMapping(MatrixWorkspace_const_sptr &workspace,
```
Comment (soininen) [inline Framework/API/inc/MantidAPI/SpectrumDetectorMapping.h]: Is it safe to pass a `shared_ptr` as a reference to a public function like this?

Comment (reimundILL) [reply Framework/API/inc/MantidAPI/SpectrumDetectorMapping.h]: Good point! I'll try to find a solution.

Comment (soininen) [reply Framework/API/inc/MantidAPI/SpectrumDetectorMapping.h]: Maybe just drop the `&` and it should be OK.

File: Framework/API/test/SpectrumDetectorMappingTest.h [hunk 2]
```diff
@@ -30,7 +32,8 @@ class SpectrumDetectorMappingTest : public CxxTest::TestSuite {
     ws->getSpectrum(0).setDetectorIDs(std::set<detid_t>());
     int detids[] = {10, 20};
     ws->getSpectrum(2).setDetectorIDs(std::set<detid_t>(detids, detids + 2));
-    SpectrumDetectorMapping map(ws.get());
+    MatrixWorkspace_const_sptr const_ws = ws->clone();
```
Comment (soininen) [inline Framework/API/test/SpectrumDetectorMappingTest.h]: Why would you need to `clone` the workspace here?

File: Framework/API/inc/MantidAPI/SpectrumDetectorMapping.h (hunk 1 above)
Comment (soininen) [inline Framework/API/inc/MantidAPI/SpectrumDetectorMapping.h]: While at it, maybe you could mark `useSpecNoIndex` as `const` as well?

**Review by soininen — CHANGES_REQUESTED** [review]

File: Framework/API/inc/MantidAPI/SpectrumDetectorMapping.h (hunk 1 above)
Comment (soininen) [reply Framework/API/inc/MantidAPI/SpectrumDetectorMapping.h]: I think you misunderstood me. I'm just suggesting changing `bool useSpecNoIndex` to `const bool useSpecNoIndex`. It has no effect on default values or initialization whatsoever.

Comment (reimundILL) [reply Framework/API/inc/MantidAPI/SpectrumDetectorMapping.h]: You're right.

**Review by soininen — APPROVED** [review]

**Pull request merged**